url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/server/genericServer/LocalStringServerSocket.java

```diff
@@ -0,0 +1,422 @@
+     * Once the server has accepted them, it will call Thread.notify on that object.
+     * 
+     * @param client Client to queue to accept
+     * @return peer Server-side peer of this client
+     *
+     * @throws IllegalStateException If we are at EOF already
+     * @throws IllegalArgumentException If client is or was accepted somewhere already
+     * @throws ConnectException If accept queue is full (ACCEPT_QUEUELENGTH)
+     * @throws EOFException  If client is at EOF already
+     *
+     * @see #accept()
+     * @see #ACCEPT_QUEUELENGTH
+     */
+    protected LocalStringConnection queueAcceptClient(LocalStringConnection client)
+        throws IllegalStateException, IllegalArgumentException, ConnectException, EOFException
+    {
+        if (isOutEOF())
+            throw new IllegalStateException("Internal error, already at EOF");
+        if (client.isAccepted())
+            throw new IllegalArgumentException("Client is already accepted somewhere");
+        if (client.isOutEOF() || client.isInEOF())
+            throw new EOFException("client is already at EOF");
+
+        // Create server-side peer of client connect object, add client
+        // to the accept queue, then notify any server thread waiting to
+        // accept clients.  Accept() callers thread-wait on the newly
+        // created peer object to prevent possible contention with other
+        // objects; we know this new object won't have any locks on it.
+
+        LocalStringConnection serverPeer = new LocalStringConnection(client);
+        synchronized (acceptQueue)
+        {
+            if (acceptQueue.size() > ACCEPT_QUEUELENGTH)
+                throw new ConnectException("Server accept queue is full");
+            acceptQueue.add(client);
+            acceptQueue.notifyAll();
+        }
+
+        return serverPeer;
+    }
+
+    /**
+     * For server to call.  Blocks waiting for next inbound connection.
+     * (Synchronizes on accept queue.)
+     * 
+     * @return The server-side peer to the inbound client connection
+     * @throws SocketException if our setEOF() has been called, thus
+     *    new clients won't receive any data from us
+     * @throws IOException if a network problem occurs (Which won't happen with this local communication)
+     */
+    public StringConnection accept() throws SocketException, IOException
+    {
+        if (out_setEOF)
+            throw new SocketException("Server socket already at EOF");
+
+        LocalStringConnection cliPeer;
+
+        synchronized (acceptQueue)
+        {
+            while (acceptQueue.isEmpty())
+            {
+                if (out_setEOF)
+                    throw new SocketException("Server socket already at EOF");
+                else
+                {
+                    try
+                    {
+                        acceptQueue.wait();  // Notified by queueAcceptClient 
+                    }
+                    catch (InterruptedException e) {}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)